### ISSUE TITLE
Generic ParamSpec on class produces `reportCallIssue` error after conditional check `len(args) == x` or `len(args) != x`

### ISSUE THREAD

**Issue by sqrtrae** [microsoft/pyright#10656]
**Describe the bug**

The below code produces the error `Arguments for ParamSpec "P@MyClass" are missing (reportCallIssue)` on the second call to the method `MyClass.func()` within `MyClass.__call__()`.

Code sample in [pyright playground](https://pyright-play.net/?code=MYGwhgzhAECyCeBhcUDaAqdAFAugLgChpjoATAUwDNoB9G4MEEOgCgnJEoBpp0wAnAOYQ80LADoBwnpgDWAdykix4hUoCU0ALQA%2BaAEsAdgBdCJc7WgBeaO07jKAV0PAWfIRBno1HzdADE0IYA9tAADvD8%2BoIAFsbQ5Pz8wfxEFsT61CDkhiwa1jYADGbpFmFRJiwA5AASHCChAOopIKQAhFXqaen85MaO-Ia2HJQOzq7u0rzeir7EgRFRsfGJyandZFTQTi5sIzJKohJKXj7CR6qzwpq6BiYlPX0DQ4UEQA)

```python
class MyClass[**P]:
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> int:
        _ = self.func(*args, **kwargs)  # no pyright error
        if len(args) == 0:
            print('Hello World!')
        return self.func(*args, **kwargs)  # pyright error

    def func(self, *args: P.args, **kwargs: P.kwargs) -> int:
        return 0
```

I did some testing with different conditions in place of `len(args) == 0` to better understand what does/doesn't produce the same error:

- Produces same error:
  - `len(args) != 0`
  - `len(args) == 1`
  - `len(args) != 1`
  - `len(args) == 314`
  - `len(args) != 314`
- Doesn't error:
  - `len(kwargs) == 0`
  - `len(args) > 10`
  - `len(args) >= 10`
  - `len(args) < 10`
  - `len(args) <= 10`

**Workaround**

This bug can be worked around by simply assigning the result of the equality check to a variable and then using that variable in the conditional, e.g.:

```python
class MyClass[**P]:
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> int:
        _ = self.func(*args, **kwargs)  # no pyright error
        comp = len(args) == 0
        if comp:
            print('Hello World!')
        return self.func(*args, **kwargs)  # no pyright error

    def func(self, *args: P.args, **kwargs: P.kwargs) -> int:
        return 0
```

**VS Code extension or command-line**

I am running pyright via the command line, using version 1.1.402.

**Comment by erictraut** [comment]
This behavior is because pyright is performing type narrowing for `args` based on the type guard expression `len(args) == 0`. Prior to narrowing, the type of `args` is a tuple of unknown length: `tuple[object, ...]`. If you check for a specific length, pyright will narrow that type accordingly. For example, if you check for `len(args) == 0)`, it will narrow the type to `tuple[()]` in the positive (if) case and `tuple[object, *tuple[object, ...]]` in the negative (else) case. 

When it performs this type narrowing, pyright loses the fact that `args` was previously related to a ParamSpec. That's why you see the error. In formal type theory, the way to indicate that a value is two different types (e.g. both `tuple[()]` _and_ ParamSpec `P`) is to use an "intersection type", sometimes denoted `&`. The Python type system doesn't currently have support for intersection types, although there is ongoing work to add it. 

Although the Python type system doesn't have support for intersections, pyright internally implements a poor-man's variant of intersections specifically to handle the case where a value whose type is described by a type variable is narrowed. I refer to this concept as a [conditional type](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables). Unfortunately for you, this capability does not currently support ParamSpecs (or TypeVarTuples). It supports only simple type variables.

I can think of a number of ways to handle this case:
1. Disable `len` type narrowing if the subject is a ParamSpec
2. Extend conditional types to support ParamSpec
3. Wait for intersections to be introduced to the type system

The first option is relatively little work and would address the issue in your code sample, but it will likely break other code that pyright users rely on, so I'm not sure if that's a good choice.

Option 2 is a lot of work — probably more than I can justify to handle this edge case since this is the first report of anyone running into this issue. Plus, this work would need to be undone when full intersections are implemented.

Option 3 is the correct long-term solution here, but this is a very heavy lift and probably won't happen for at least another year.

I'll think about this a bit more, but my current inclination is to prototype option 1 and run that against some existing code bases to understand the impact. If it doesn't appear to have a significant impact on other code bases, then I think it's a reasonable short-term fix. If it has a big impact on other code bases, then I think option 3 is the right answer.

**Comment by sqrtrae** [comment]
Thanks for the insight into how pyright is evaluating this scenario! I didn't realize the type of `args` changes like this by the time it is used in the second call to `MyClass.func()`.

I don't have much to opine on in terms of how you want to fix this (or if you do want to). In my own case, the workaround I gave is easy enough for me to use instead, despite being slightly less pretty.
